addons/account/models/account_move.py

```diff
@@ -1491,11 +1491,22 @@ def _compute_amount(self):
             if new_pmt_state == 'paid' and move.move_type in ('in_invoice', 'out_invoice', 'entry'):
                 reverse_type = move.move_type == 'in_invoice' and 'in_refund' or move.move_type == 'out_invoice' and 'out_refund' or 'entry'
                 reverse_moves = self.env['account.move'].search([('reversed_entry_id', '=', move.id), ('state', '=', 'posted'), ('move_type', '=', reverse_type)])
-
-                # We only set 'reversed' state in cas of 1 to 1 full reconciliation with a reverse entry; otherwise, we use the regular 'paid' state
-                reverse_moves_full_recs = reverse_moves.mapped('line_ids.full_reconcile_id')
-                if reverse_moves_full_recs.mapped('reconciled_line_ids.move_id').filtered(lambda x: x not in (reverse_moves + reverse_moves_full_recs.mapped('exchange_move_id'))) == move:
-                    new_pmt_state = 'reversed'
+                if reverse_moves:
+                    # We only set 'reversed' state in cas of 1 to 1 full reconciliation with a reverse entry; otherwise, we use the regular 'paid' state
+                    if move.is_invoice():
+                        current_lines = move.line_ids.filtered(lambda line: line.account_id.user_type_id.type in ('receivable', 'payable'))
+                        target_lines = reverse_moves.line_ids.filtered(lambda line: line.account_id.user_type_id.type in ('receivable', 'payable'))
+                    else:
+                        current_lines = move.line_ids
+                        target_lines = reverse_moves.line_ids
+                    reverse_moves_full_recs = target_lines.full_reconcile_id
+                    excluded_moves = set(
+                        (current_lines.matched_debit_ids + current_lines.matched_credit_ids).exchange_move_id
+                        + reverse_moves_full_recs.exchange_move_id
+                        + reverse_moves
+                    )
+                    if reverse_moves_full_recs.reconciled_line_ids.move_id.filtered(lambda x: x not in excluded_moves) == move:
+                        new_pmt_state = 'reversed'
```

Comment:
that code should be rewritten in a clearer way (the old code was crappy, not your fault, but still)

* Comment that you need to exclude the exchange move(s)
* maybe simply check the length == 2 and composed as the recordset(move, reverse_moves[0]) ?

Reply:
All this code should be made in SQL. I was terrified when I read this part of the code... this is a disaster regarding the performances